### PR TITLE
[MB-11010] PPM booking dates and location form storybook

### PR DESCRIPTION
**Pull request by duncan-truss**
## Summary

This PR creates the new DatesAndLocations.jsx form component for the PPM Booking Dates and Location page and adds it to storybook.

The component has conditional fields displayed for the secondary ZIPs, these become required when the has additional ZIP questions are set to "Yes".

The origin ZIP can be prefilled with the residential address if the checkbox is selected.  The destination ZIP can be prefilled with the destination duty location ZIP if the checkbox is selected.

The ZIP codes have an async validation to check if they are valid according to the rate engine.  This is an existing endpoint in the Internal API that is called from elsewhere in the Customer app.  It checks different tables depending on if it's the origin ZIP or the destination, not sure if this behavior is still desired.

We also talked about possibly disabling the ZIP text boxes if the checkbox to use the residential address or destination duty location is chosen so we don't send a payload to the API that says we're using an existing one but then the user has edited the field.  This may not matter too much for the estimate but something to consider about weird states.

[Zeplin hi-fi mocks](https://zpl.io/g8qnW8Z)

TODO

- [x] Add section styling
- [x] Add mobile layout styling
- [x] Add button styling
- [x] Add additional storybook stories (existing PPM, optional fields, validations)
- [x] Fill set origin/destination ZIP values (residential ZIP, destination duty station ZIP) when checkboxes are selected
- [x] React Testing Library component unit tests

[this article](tbd) explains more about the approach used.

## Setup to Run Your Code

<details>
<summary>💻 You will need to use three separate terminals to test this locally.</summary>

##### Terminal 1

Start the Storybook locally.

```sh
make storybook
```

##### Terminal 2

Start the UI locally.

```sh
make client_run
```
</details>

### Additional steps

1. Open storybook locally and find the Customer Components -> PPM Booking -> Dates And Locations component http://localhost:6006/?path=/story/customer-components-ppm-booking-dates-and-location--blank-dates-and-location

## Verification Steps for Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Request review from a member of a different team.
- [ ] Have the Jira acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Frontend

- [ ] User facing changes have been reviewed by design.
- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
- [ ] There are no new console errors in the browser devtools
- [ ] There are no new console errors in the test output
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/dev/contributing/backend/Backend-Programming-Guide#logging)
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/dev/contributing/backend/Backend-Programming-Guide/#querying-the-database-safely) have been satisfied.

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/dev/contributing/database/Database-Migrations#zero-downtime-migrations)
- [ ] Have been communicated to #g-database
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/dev/contributing/database/Database-Migrations#secure-migrations)

## Screenshots
Screen Shot 2022-02-08 at 12 00 49 PM
Screen Shot 2022-02-08 at 12 01 30 PM
Screen Shot 2022-02-08 at 12 02 56 PM
Screen Shot 2022-02-08 at 12 04 03 PM

### CODE REVIEW

File: src/components/Customer/PPMBooking/DatesAndLocation/DatesAndLocation.stories.jsx [hunk 1]
```diff
@@ -1,23 +1,12 @@
...
 import { action } from '@storybook/addon-actions';
-import { Grid, GridContainer } from '@trussworks/react-uswds';
+import { within, userEvent } from '@storybook/testing-library';
 
 import DatesAndLocation from './DatesAndLocation';
 
 export default {
   title: 'Customer Components / PPM Booking / Dates and Location',
   component: DatesAndLocation,
-  decorators: [
```
Comment (duncan-truss) [inline src/components/Customer/PPMBooking/DatesAndLocation/DatesAndLocation.stories.jsx]: Did you remove this for a reason or did it just get lost in the commits?

**Comment by duncan-truss** [comment]
@dinabeigelman You should be able to view this PR's storybook at this URL https://598411-114694829-gh.circle-artifacts.com/0/storybook/index.html?path=/story/customer-components-ppm-booking-dates-and-location--blank-dates-and-location

There are 6 new Happo diffs (2 flakey unrelated diffs) you can review here https://happo.io/a/488/p/603/compare/235b98dc2b65acc1a61f3fdaa74d11caab61fb1a/d2b01f0a7518e53a6311aa94fea07a4bc5ce665a

File: src/components/Customer/PPMBooking/DatesAndLocation/DatesAndLocation.test.jsx [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    residentialAddress: {
+      postalCode: '90210',
+    },
+  },
+  destinationDutyStation: {
+    address: {
+      postalCode: '94611',
+    },
+  },
+  postalCodeValidator: jest.fn(),
```
Comment (ronaktruss) [inline src/components/Customer/PPMBooking/DatesAndLocation/DatesAndLocation.test.jsx:38]: Might be worth replacing with:
```suggestion
  ...defaultProps,
```

**Review by ronaktruss — APPROVED** [review]
LGTM outside of existing comments

**Comment by scarequotes** [comment]
Recommended for a later story: 

When a customer checks "use my existing ZIP", then edits the ZIP field manually, the checkbox for "use existing" should automatically un-select. 

This avoids scenarios where a customer types a different ZIP but the "use my current ZIP (96766)" field is active/checked and shows a different ZIP code. 

(What we have now matches what happens already in HHG and other shipments, so this change should be consistent throughout.)

**Comment by scarequotes** [comment]
ZIP validation message should be `Enter a 5-digit ZIP code` (should be true everywhere in the onboarding/booking experience, not just PPM)

**Comment by scarequotes** [comment]
A 9-digit ZIP without a hyphen should not show an error — ideally would just autoformat to add the hyphen `XXXXX-XXXX`

**Review by scarequotes — APPROVED** [review]
Left some comments about validation issues and error messaging that should be fixed beyond the scope of this PR, but this is good to go.

**Pull request merged**